ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs

```diff
@@ -36,32 +36,37 @@ protected ReactContextNativeModuleBase(ReactContext reactContext, IActionQueue a
         }
 
         /// <summary>
-        /// Instantiates the <see cref="ReactContextNativeModuleBase"/>.
+        /// The action queue used by the native module.
         /// </summary>
-        /// <param name="reactContext">The React context.</param>
-        /// <param name="delegateFactory">The React method delegate factory.</param>
-        protected ReactContextNativeModuleBase(ReactContext reactContext, IReactDelegateFactory delegateFactory)
-            : base(delegateFactory)
+        /// <remarks>
+        /// Defaults to the native modules action queue. If you override this getter,
+        /// be sure to return the same IActionQueue instance each time this getter is
+        /// accessed for a given module instance.
+        /// </remarks>
+        public virtual IActionQueue ActionQueue
         {
-            if (reactContext == null)
-                throw new ArgumentNullException(nameof(reactContext));
-
-            Context = reactContext;
+            get
+            {
+                return Context.IndisposableNativeModulesQueue;
+            }
```

Comment:
What was the reason why we couldn't just provide a public interface (e.g., `IHasActionQueue`) and in the `NativeModuleRegistry` do a type check before dispatching the action?

Reply:
If the user wants to enqueue on the `NativeModulesActionQueue` they can do so by interacting through the `ReactContext`...

---
In reply to: [174975401](https://github.com/Microsoft/react-native-windows/pull/1650#discussion_r174975401) [](ancestors = 174975401)

Reply:
Such an approach wouldn't limit us to only dispatching through the `ReactContextNativeModuleBase`, but any `INativeModule` that is also `IHasActionQueue`.

---
In reply to: [174975466](https://github.com/Microsoft/react-native-windows/pull/1650#discussion_r174975466) [](ancestors = 174975466,174975401)

Reply:
`NativeModuleRegistry` does a type check for `IHasActionQueue`.

We could make `IHasActionQueue` public so that people can add it to any `INativeModule`.

The current design is we have the `IHasActionQueue` interface and only `ReactContextNativeModuleBase` implements it by default. If you do not implement `IHasActionQueue`, you run on the native modules queue. If you do implement `IHasActionQueue`, then `ActionQueue` must be non-null and you run on that queue.

A complication that lead to this design is that I want to enforce that `ActionQueue` is non-null. The default `ActionQueue` is the native modules queue but `NativeModuleBase` doesn't have any way to reference the native modules queue. The only ways we thought of to give `NativeModuleBase` access to the native modules queue involved a breaking change or a global variable.

It's true that a module can enqueue to `NativeModulesActionQueue ` through `ReactContext` but I prefer it be done through `ActionQueue`. This gives module authors a consistent way to put work onto the module's queue regardless of what the queue is. Also, if you later change the module's queue you don't have to update all of the module's queueing code.

---
In reply to: [174975581](https://github.com/Microsoft/react-native-windows/pull/1650#discussion_r174975581) [](ancestors = 174975581,174975466,174975401)